File: apps/api/src/modules/auth/auth.service.ts
import { Injectable, UnauthorizedException, BadRequestException, InternalServerErrorException, Logger } from '@nestjs/common';
import { FirebaseService } from '../../common/firebase/firebase.service';
import { AuthRepository } from './auth.repository';
import { CurrentUserPayload } from '../../common/decorators/current-user.decorator';
import { RegisterDto } from './dto/register.dto';
import { RegisterFcmDto } from './dto/register-fcm.dto';

// Simple in-memory OTP store (use Redis in production)
const OTP_STORE = new Map<string, { otp: string; expiresAt: number }>();
const OTP_TTL_MS = 5 * 60 * 1000; // 5 minutes

function generateOtp(): string {
  if (process.env.NODE_ENV !== 'production') return '999999';
  return String(Math.floor(100000 + Math.random() * 900000));
}

function normalizePhone(phone: string): string {
  // Ensure +84 format for Vietnam
  const digits = phone.replace(/\D/g, '');
  // Handle +840xxxxxxx (frontend sends dialCode+localWithLeadingZero)
  if (digits.startsWith('840')) return `+84${digits.slice(3)}`;
  if (digits.startsWith('84')) return `+${digits}`;
  if (digits.startsWith('0')) return `+84${digits.slice(1)}`;
  return `+${digits}`;
}

@Injectable()
export class AuthService {
  private readonly logger = new Logger(AuthService.name);

  constructor(
    private readonly firebase: FirebaseService,
    private readonly repo: AuthRepository,
  ) {}

  async verifyAndGetOrCreateUser(idToken: string) {
    let decoded: Awaited<ReturnType<typeof this.firebase.verifyIdToken>>;
    try {
      decoded = await this.firebase.verifyIdToken(idToken);
    } catch (err: unknown) {
      const code = (err as { code?: string })?.code ?? '';
      this.logger.error(`verifyIdToken failed: code=${code}`, err);
      if (code === 'auth/id-token-expired') throw new UnauthorizedException('토큰이 만료되었습니다.');
      if (code === 'auth/argument-error' || code === 'auth/invalid-id-token') throw new UnauthorizedException('유효하지 않은 토큰입니다.');
      throw new UnauthorizedException(`Firebase 인증 실패: ${code || String(err)}`);
    }

    // 1. Try to find by firebase_uid (existing user)
    const existing = await this.repo.findByFirebaseUid(decoded.uid);
    if (existing) {
      // Backfill email and name from Google/social provider if not yet set
      if (decoded.email && !existing.email) {
        await this.repo.updateEmailIfNull(existing.id, decoded.email);
      }
      if (decoded.name) {
        await this.repo.updateWorkerNameIfNull(existing.id, decoded.name);
      }
      const profile = await this.repo.getMeProfile(existing.id);
      return { user: profile, isNew: false };
    }

    // 2. If phone is present, check if there's already a user with that phone
    //    (can happen when Firebase account is recreated after deletion)
    if (decoded.phone_number) {
      const byPhone = await this.repo.findByPhone(decoded.phone_number);
      if (byPhone) {
        // Link new Firebase UID to the existing account
        await this.repo.updateFirebaseUid(byPhone.id, decoded.uid);
        const profile = await this.repo.getMeProfile(byPhone.id);
        return { user: profile, isNew: false };
      }
    }

    // 3. Create new user
    const providerMap: Record<string, string> = {
      'google.com': 'google',
      'facebook.com': 'facebook',
      'phone': 'phone',
      'password': 'email',
    };
    const provider = providerMap[decoded.firebase?.sign_in_provider ?? 'phone'] ?? 'phone';

    try {
      const user = await this.repo.create({
        firebaseUid: decoded.uid,
        phone: decoded.phone_number || null,
        email: decoded.email || null,
        role: 'WORKER',
        provider,
      });
      await this.repo.ensureWorkerProfile(user.id, decoded.phone_number || null, decoded.name || null);
      const profile = await this.repo.getMeProfile(user.id);
      return { user: profile, isNew: true };
    } catch (err: unknown) {
      const pgErr = err as { code?: string; constraint?: string };
      this.logger.error(`DB error creating user uid=${decoded.uid}`, err);
      if (pgErr.code === '23505') {
        // Unique violation — race condition or duplicate; find the existing user
        const fallback = await this.repo.findByFirebaseUid(decoded.uid);
        if (fallback) {
          const profile = await this.repo.getMeProfile(fallback.id);
          return { user: profile, isNew: false };
        }
      }
      throw new InternalServerErrorException('사용자 생성에 실패했습니다.');
    }
  }

  async register(currentUser: CurrentUserPayload, dto: RegisterDto) {
    return this.repo.updateRole(currentUser.id, dto.role);
  }

  async updateProfile(userId: string, data: { name?: string; email?: string; password?: string }) {
    // If email or password provided, update Firebase user as well
    if (data.email || data.password) {
      const dbUser = await this.repo.findById(userId);
      if (dbUser) {
        await this.firebase.updateFirebaseUser(dbUser.firebase_uid, {
          email: data.email,
          password: data.password,
        });
      }
    }
    return this.repo.updateProfile(userId, { name: data.name, email: data.email });
  }

  async linkPhone(userId: string, phoneIdToken: string): Promise<void> {
    const decoded = await this.firebase.verifyIdToken(phoneIdToken);
    if (!decoded.phone_number) {
      throw new BadRequestException('Phone ID token does not contain a phone number');
    }
    await this.repo.updatePhone(userId, decoded.phone_number);
  }

  async getMe(userId: string) {
    return this.repo.getMeProfile(userId);
  }

  async registerFcmToken(userId: string, dto: RegisterFcmDto) {
    return this.repo.upsertFcmToken(userId, dto.token, dto.platform);
  }

  async isTestPhone(phone: string): Promise<{ isTest: boolean }> {
    const isTest = await this.repo.isTestPhone(phone);
    return { isTest };
  }

  // ── OTP flow ────────────────────────────────────────────────────────────────

  async sendOtp(phone: string): Promise<{ message: string; devOtp?: string }> {
    const normalized = normalizePhone(phone);
    const otp = generateOtp();
    OTP_STORE.set(normalized, { otp, expiresAt: Date.now() + OTP_TTL_MS });

    // In production: send SMS via your SMS provider here
    // For now: log to console + return in dev mode
    console.log(`[OTP] ${normalized}: ${otp}`);

    const isDev = process.env.NODE_ENV !== 'production';
    return {
      message: '인증번호가 발송되었습니다.',
      ...(isDev ? { devOtp: otp } : {}),
    };
  }

  async verifyOtp(phone: string, otp: string): Promise<{ customToken?: string; devToken?: string; isNewUser: boolean }> {
    const normalized = normalizePhone(phone);
    const stored = OTP_STORE.get(normalized);

    if (!stored || stored.expiresAt < Date.now()) {
      throw new UnauthorizedException('인증번호가 만료되었습니다.');
    }
    if (stored.otp !== otp) {
      throw new UnauthorizedException('인증번호가 올바르지 않습니다.');
    }

    OTP_STORE.delete(normalized);

    // Find or create Firebase user by phone
    const { uid, isNew: isFirebaseNew } = await this.firebase.getOrCreateUserByPhone(normalized);

    // Find or create DB user (try by firebase uid first, then phone as fallback)
    let dbUser = await this.repo.findByFirebaseUid(uid);
    let isNewUser = isFirebaseNew;
    if (!dbUser) {
      dbUser = await this.repo.findByPhone(normalized);
      if (dbUser) {
        // Phone exists but with a different firebase uid — update it
        await this.repo.updateFirebaseUid(dbUser.id, uid);
        dbUser = { ...dbUser, firebase_uid: uid };
      } else {
        dbUser = await this.repo.create({ firebaseUid: uid, phone: normalized, email: null, role: 'WORKER' });
        await this.repo.ensureWorkerProfile(dbUser.id, normalized);
        isNewUser = true;
      }
    }

    const isDev = process.env.NODE_ENV !== 'production';
    if (isDev) {
      // Dev mode: return a simple token format (dev_<userId>)
      return { devToken: `dev_${dbUser.id}`, isNewUser };
    }

    const customToken = await this.firebase.createCustomToken(uid);
    return { customToken, isNewUser };
  }

  // ── Email + password flow ────────────────────────────────────────────────────

  async loginEmail(email: string, password: string): Promise<{ customToken?: string; devToken?: string }> {
    // Use Firebase REST API to verify email/password
    const apiKey = process.env.FIREBASE_WEB_API_KEY || process.env.NEXT_PUBLIC_FIREBASE_API_KEY || '';

    if (!apiKey) {
      throw new BadRequestException('Email login not configured');
    }

    const res = await fetch(
      `https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=${apiKey}`,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ email, password, returnSecureToken: true }),
      },
    );

    if (!res.ok) {
      throw new UnauthorizedException('이메일 또는 비밀번호가 올바르지 않습니다.');
    }

    const data = await res.json() as { localId: string };
    const uid = data.localId;

    // Ensure user exists in DB
    let dbUser = await this.repo.findByFirebaseUid(uid);
    if (!dbUser) {
      dbUser = await this.repo.create({ firebaseUid: uid, phone: null, email, role: 'WORKER' });
    }

    const isDev = process.env.NODE_ENV !== 'production';
    if (isDev) {
      return { devToken: `dev_${dbUser.id}` };
    }

    const customToken = await this.firebase.createCustomToken(uid);
    return { customToken };
  }

  // ── Facebook social login ────────────────────────────────────────────────────

  async socialFacebook(idToken: string): Promise<{ isNewUser: boolean; needsPhone: boolean; devToken?: string }> {
    if (!idToken) throw new UnauthorizedException('idToken이 필요합니다.');
    let decoded: Awaited<ReturnType<typeof this.firebase.verifyIdToken>>;
    try {
      decoded = await this.firebase.verifyIdToken(idToken);
    } catch {
      throw new UnauthorizedException('유효하지 않은 Firebase ID Token입니다.');
    }
    let dbUser = await this.repo.findByFirebaseUid(decoded.uid);
    let isNewUser = false;

    if (!dbUser) {
      dbUser = await this.repo.create({
        firebaseUid: decoded.uid,
        phone: decoded.phone_number || null,
        email: decoded.email || null,
        role: 'WORKER',
        provider: 'facebook',
      });
      await this.repo.ensureWorkerProfile(dbUser.id, decoded.phone_number || null, decoded.name || null);
      isNewUser = true;
    }

    const needsPhone = !dbUser.phone;
    const isDev = process.env.NODE_ENV !== 'production';
    return {
      isNewUser,
      needsPhone,
      ...(isDev ? { devToken: `dev_${dbUser.id}` } : {}),
    };
  }

  // ── Logout ───────────────────────────────────────────────────────────────────

  async logout(firebaseUid: string): Promise<void> {
    try {
      await this.firebase.revokeRefreshTokens(firebaseUid);
    } catch {
      // Ignore — local session will be cleared by client
    }
  }
}

File: apps/api-kotlin/src/main/kotlin/vn/gada/api/auth/AuthService.kt
package vn.gada.api.auth

import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import vn.gada.api.common.exception.BadRequestException
import vn.gada.api.common.exception.UnauthorizedException
import vn.gada.api.common.firebase.FirebaseService

@Service
class AuthService(
    private val repo: AuthRepository,
    private val firebase: FirebaseService,
    private val otpStore: OtpStore,
    @Value("\${spring.profiles.active:default}") private val activeProfile: String,
    @Value("\${gada.firebase.web-api-key:}") private val firebaseWebApiKey: String
) {
    private val log = LoggerFactory.getLogger(AuthService::class.java)
    private val isDev: Boolean get() = activeProfile != "production" && activeProfile != "prod"

    fun verifyAndGetOrCreateUser(idToken: String, name: String? = null, emailOverride: String? = null): Map<String, Any?> {
        val decoded = firebase.verifyIdToken(idToken)
            ?: throw UnauthorizedException("Invalid token")
        val existing = repo.findByFirebaseUid(decoded.uid)
        if (existing != null) {
            // If account was soft-deleted, treat as new registration
            if (existing["status"] == "DELETED") {
                repo.reactivateUser(existing["id"] as String, name ?: (decoded.claims["phone_number"] as? String))
                val reactivated = repo.findByFirebaseUid(decoded.uid)!!
                return mapOf("user" to reactivated, "isNew" to true)
            }
            // Ensure worker_profile exists (handles incomplete registrations)
            val userId = existing["id"] as String
            repo.ensureWorkerProfile(userId, name ?: (existing["phone"] as? String))
            return mapOf("user" to existing, "isNew" to false)
        }
        val phoneNumber = decoded.claims["phone_number"] as? String
        val tokenEmail = decoded.claims["email"] as? String
        val email = emailOverride ?: tokenEmail
        val user = repo.create(decoded.uid, phoneNumber, email, "WORKER")
        repo.ensureWorkerProfile(user["id"] as String, name ?: phoneNumber)
        return mapOf("user" to user, "isNew" to true)
    }

    fun getMe(userId: String): Map<String, Any?>? {
        return repo.getMeProfile(userId)
    }

    fun updateProfile(userId: String, name: String?, email: String?): Map<String, Any?>? {
        return repo.updateProfile(userId, name, email)
    }

    fun registerFcmToken(userId: String, token: String, platform: String): Map<String, Any> {
        repo.upsertFcmToken(userId, token, platform)
        return mapOf("success" to true)
    }

    // ── OTP flow ─────────────────────────────────────────────────────────────

    fun sendOtp(phone: String): Map<String, Any?> {
        val normalized = normalizePhone(phone)
        val otp = otpStore.generateOtp()
        otpStore.save(normalized, otp)

        log.info("[OTP] {}: {}", normalized, otp)

        return if (isDev) {
            mapOf("message" to "인증번호가 발송되었습니다.", "devOtp" to otp)
        } else {
            mapOf("message" to "인증번호가 발송되었습니다.")
        }
    }

    fun verifyOtp(phone: String, otp: String): Map<String, Any?> {
        val normalized = normalizePhone(phone)

        if (otpStore.isExpiredOrMissing(normalized)) {
            throw UnauthorizedException("인증번호가 만료되었습니다.")
        }

        val valid = otpStore.verify(normalized, otp)
        if (!valid) {
            throw UnauthorizedException("인증번호가 올바르지 않습니다.")
        }

        // Find or create Firebase user by phone
        val (uid, isFirebaseNew) = firebase.getOrCreateUserByPhone(normalized)

        // Find or create DB user
        var dbUser = repo.findByFirebaseUid(uid)
        var isNewUser = isFirebaseNew
        if (dbUser == null) {
            dbUser = repo.findByPhone(normalized)
            if (dbUser != null) {
                // Phone exists but different firebase uid — update it
                repo.updateFirebaseUid(dbUser["id"] as String, uid)
                dbUser = dbUser.toMutableMap().apply { put("firebase_uid", uid) }
            } else {
                dbUser = repo.create(uid, normalized, null, "WORKER")
                repo.ensureWorkerProfile(dbUser["id"] as String, normalized)
                isNewUser = true
            }
        }

        return if (isDev) {
            mapOf("devToken" to "dev_${dbUser["id"]}", "isNewUser" to isNewUser)
        } else {
            val customToken = firebase.createCustomToken(uid)
            mapOf("customToken" to customToken, "isNewUser" to isNewUser)
        }
    }

    // ── Email + password login ────────────────────────────────────────────────

    fun loginEmail(email: String, password: String): Map<String, Any?> {
        if (firebaseWebApiKey.isBlank()) {
            throw BadRequestException("Email login not configured")
        }

        val url = "https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key=$firebaseWebApiKey"
        val body = """{"email":"$email","password":"$password","returnSecureToken":true}"""

        val connection = java.net.URL(url).openConnection() as java.net.HttpURLConnection
        connection.requestMethod = "POST"
        connection.setRequestProperty("Content-Type", "application/json")
        connection.doOutput = true
        connection.outputStream.write(body.toByteArray())

        if (connection.responseCode != 200) {
            throw UnauthorizedException("이메일 또는 비밀번호가 올바르지 않습니다.")
        }

        val responseBody = connection.inputStream.bufferedReader().readText()
        val uid = extractJsonField(responseBody, "localId")
            ?: throw UnauthorizedException("이메일 또는 비밀번호가 올바르지 않습니다.")

        var dbUser = repo.findByFirebaseUid(uid)
        if (dbUser == null) {
            dbUser = repo.create(uid, null, email, "WORKER")
            repo.ensureWorkerProfile(dbUser["id"] as String, null)
        }

        return if (isDev) {
            mapOf("devToken" to "dev_${dbUser["id"]}")
        } else {
            val customToken = firebase.createCustomToken(uid)
            mapOf("customToken" to customToken)
        }
    }

    // ── Facebook social login ─────────────────────────────────────────────────

    fun socialFacebook(idToken: String, name: String? = null, emailOverride: String? = null): Map<String, Any?> {
        val decoded = firebase.verifyIdToken(idToken)
            ?: throw UnauthorizedException("Invalid token")
        var dbUser = repo.findByFirebaseUid(decoded.uid)
        var isNewUser = false
        if (dbUser == null) {
            val phoneNumber = decoded.claims["phone_number"] as? String
            val tokenEmail = decoded.claims["email"] as? String
            val email = emailOverride ?: tokenEmail
            dbUser = repo.create(decoded.uid, phoneNumber, email, "WORKER")
            repo.ensureWorkerProfile(dbUser["id"] as String, name ?: phoneNumber)
            isNewUser = true
        } else {
            // Ensure worker_profile has a name (handles incomplete SSO registrations)
            if (name != null) {
                repo.ensureWorkerProfile(dbUser["id"] as String, name)
            }
        }
        return mapOf("isNewUser" to isNewUser)
    }

    // ── Test account login (staging only) ────────────────────────────────────

    fun testLogin(role: String): Map<String, Any?> {
        if (!isDev) throw BadRequestException("Not available in production")
        val normalizedRole = role.uppercase()
        if (normalizedRole != "WORKER" && normalizedRole != "MANAGER") {
            throw BadRequestException("role must be WORKER or MANAGER")
        }
        val user = repo.upsertTestAccount(normalizedRole)
        return mapOf("devToken" to "dev_${user["id"]}", "role" to normalizedRole)
    }

    // ── Logout ────────────────────────────────────────────────────────────────

    fun logout(firebaseUid: String) {
        try {
            firebase.revokeRefreshTokens(firebaseUid)
        } catch (e: Exception) {
            // Ignore — local session will be cleared by client
            log.debug("Token revocation failed for {}: {}", firebaseUid, e.message)
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    fun normalizePhone(phone: String): String {
        val digits = phone.replace(Regex("[^\\d]"), "")
        return when {
            digits.startsWith("840") -> "+84${digits.drop(3)}"
            digits.startsWith("84") -> "+$digits"
            digits.startsWith("0") -> "+84${digits.drop(1)}"
            else -> "+$digits"
        }
    }

    private fun extractJsonField(json: String, field: String): String? {
        val pattern = Regex(""""$field"\s*:\s*"([^"]+)"""")
        return pattern.find(json)?.groupValues?.get(1)
    }
}
